feat: enable receipt download

tshuli opened the pull request:
## Problem
- Previously, after payment is completed, we provide a link for user to view the payment on Stripe website
- This PR introduces a new payment receipt download endpoint which returns the receipt pdf. The receipt is then downloaded onto the client machine.
- Closes #5808

**Details**
We make use of the [Content-Disposition](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) header to download the file directly using the browser. 

**Screenshot**
- Receipt is saved as pdf
![chrome-capture-2023-2-13](https://user-images.githubusercontent.com/56983748/224651770-558049be-9b6a-4541-9d6f-d479e308cb41.gif)


## Tests
- [ ] Create a form with payment. Complete payment. You should see receipt download button. Test that receipt download works successfully.

tshuli committed:
feat: popup modal to confirm payment (#5829)

Commit:
feat: set up receipt status endpoint

Commit:
feat: download receipt endpoint which returns pdf

Commit:
chore: use mutation for download

timotheeg commented on frontend/src/features/public-form/PublicFormService.ts:
```diff
@@ -152,16 +154,37 @@ export const submitFormFeedback = async (
+ * @param submissionId the id of the form submission
+ * @returns void
+ */
+export const downloadPaymentReceipt = async (
+  formId: string,
+  submissionId: string,
+): Promise<void> => {
+  return ApiService.get(
+    `payments/receipt/${formId}/${submissionId}/download`,
+  ).then(({ data }) => {
+    const buffer = Buffer.from(data.pdfBuffer.data)
+
+    const blob = new Blob([buffer], { type: 'application/pdf' })
+    FileSaver.saveAs(blob, `${submissionId}-receipt.pdf`)
+    return
+  })
+}
```
I don't think we need to handle this ourselves. We can leverage the browser's native capabilities to download a file by using the [Content-Disposition header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) and not have to hold anything in memory. The `Download` button should just act as link to make it happen.

There is additional an issue with separation of concerns here 🤔. The file is generated as pdf on the server, so the server should set a content-type header and the appropriate file name+extension in the response.

timotheeg commented on src/app/utils/convert-html-to-pdf.ts:
```diff
@@ -0,0 +1,33 @@
+export const generatePdfFromHtml = async (
+  summaryHtml: string,
+): Promise<Buffer> => {
+  const browser = await puppeteer.launch({
```
Let's wrap this function with a DD span, so we can monitor how long the pdf generation takes.

wanlingt replied:
[a1e656d](https://github.com/opengovsg/FormSG/pull/5837/commits/a1e656d1de54860d037c5e4591a7c0470c22fff9)

timotheeg commented on src/app/modules/payments/stripe.controller.ts:
```diff
@@ -191,17 +193,70 @@ export const getPaymentReceipt: ControllerHandler<{
+            return pdfBufferPromise
+          })
+          .then((pdfBuffer) => {
+            return res.status(StatusCodes.OK).json({ pdfBuffer })
```
I don't think json is the right medium to send a full file content. Let's have this endpoint return the file as-is, And supply relevant headers:
* [Content-Type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type)
* [Content-Disposition](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition)

Also [Content-Length](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Length) if express doesn't include that by default.

wanlingt replied:
[9a2c271](https://github.com/opengovsg/FormSG/pull/5837/commits/9a2c2716692a8899b0f75ce1f9bd3f0fca1cc002)

Content-Length is automatically assigned by express-js ([source](http://expressjs.com/en/4x/api.html))

timotheeg reviewed (commented):
Good work!

I'm not really a fan of us taking over pdf rendering when the browser has this capability built-in 🤔 . This case is unlike the pdf attachment to emails, where the whole operation is server-side. 

And even if we do take over pdf generation, we should leverage browser capabilities wherever we can. Attachment download in binary is well supported, so there's no need for us to transmit a binary file in a json envelop, encoded to fit in a string and taking more space than needed in transmission.

wanlingt committed:
feat: add some payment details to responses page (#5839)

* feat: add payment status

* feat: use dollars for payment amount

* feat: add created property to payments

* fix: add conditional for paymentData

* fix: use en-GB

* fix: add timezone and submission time to indiv responses page

* fix: revert payoutDate to use Date format

* feat: make payment status uppercase

foochifa committed:
feat: implement payment beta (#5831)

* feat: add script to add payment betaflag by email

* feat: update user schema with payment betaflag

* feat: remove rendering of payment settings and sidebar for non payment beta users

* feat: add utility function to verify if user has selected betaflag

* feat: prevent user from accessing payment and stripe apis if beta flag is not enabled

* fix: typo bug in payments flag in user.server.model

foochifa committed:
fix: decouple payment response from individual response page (#5861)

* fix: add individual payment response component

* fix: abstracted payment component away from individual response page

* fix: utilise skeleton loading in individual response page for payment response

* feat: add loading and error state for IndividualPaymentResponse

---------

Co-authored-by: wanlingt <wanling@open.gov.sg>

LinHuiqing committed:
feat: input and display payment amount in [dollars].[cents] (#5852)

* feat: display payment amount in 2dp

* fix: floating point error due to multiplication

example input that causes this: 34289.73

* feat: add cosmetic changes to align with design

* fix: borderLeftRadius = 0 only if there's a prefix

* refactor: separate MoneyInput from NumberInput

* test: moneyInput unit tests

* fix: test input specific UI features

removed SGD hardcode

* refactor: typing for payment amount and validation

- use FormPayments for internal state, FormPaymentsDeisplay for form fields state
- validate on MoneyInput instead of mutating value to fit requirements

* refactor: clean destructuring for common payment keys

* fix: do string conversion to and from cents

* fix: useMemo currencyFormatter complaint

* refactor: extract shared payment functions to utils

- use cents-dollars converters on PaymentPageBlock + PaymentContent too

* refactor: use of spread operator in objects

foochifa committed:
feat: shift payment acc id details from drawer to settings (#5903)

* feat: add account id to payments settings page

* fix: optimise the field to utilise FormLabel element

* feat: add account_id as a parameter to

* fix: remove target_account_id from payment drawer

KenLSM committed:
feat: only allow card payment from stripe

foochifa committed:
fix: remove paynow and grabpay from PaymentMock.png to mirror respondent side

wanlingt committed:
feat: add tracer span for generatePdfFromHtml

wanlingt committed:
fix: set express headers when returning pdf buffer

wanlingt committed:
fix: set filename in header

wanlingt committed:
fix: use browser's inbuilt save as function

wanlingt committed:
fix: return data in downloadPaymentReceipt

wanlingt committed:
fix: fix conflicts

wanlingt committed:
fix: open receipt download in new tab

wanlingt committed:
feat: set up proxy middleware for local dev

wanlingt committed:
fix: remove mutation

wanlingt committed:
Merge branch 'feat/payment-mvp' into feat/payment-receipt-download

timotheeg commented on frontend/src/setupProxy.js:
```diff
@@ -0,0 +1,15 @@
+/* eslint-disable no-undef */
+const { createProxyMiddleware } = require('http-proxy-middleware')
+
+// Added a proxy middleware as api routes with text/html in its Accept header
+// were not getting redirected to the proxy
```
Clarification on the behaviour (+ GammarNazi 😜)

```suggestion
// Added a proxy middleware, as requests to api routes with text/html in their Accept header
// were not getting proxied to the backend
```

wanlingt replied:
hahaha thanks tim, [aa46d3c](https://github.com/opengovsg/FormSG/pull/5837/commits/aa46d3c02b7afecbf6c75c33cca9b9bfb545f0b1)

timotheeg commented on src/app/modules/payments/stripe.controller.ts:
```diff
@@ -191,17 +193,76 @@ export const getPaymentReceipt: ControllerHandler<{
+          // convert to pdf and return
+          .then((receiptUrlResponse) => {
+            const html = receiptUrlResponse.data
+            return tracer.trace('generatePdfFromHtml', () => {
+              const pdfBufferPromise = generatePdfFromHtml(html)
+              return pdfBufferPromise
+            })
```
It's probably better to create the DD span in the function `generatePdfFromHtml()` itself, so it will create spans regardless of who the caller is.

wanlingt replied:
[e8982e6](https://github.com/opengovsg/FormSG/pull/5837/commits/e8982e6dabba323362c014aae8d95ab16fa3dd61)

timotheeg commented on src/app/routes/api/v3/payments/payments.routes.ts:
```diff
@@ -7,14 +7,25 @@ export const PaymentsRouter = Router()
 PaymentsRouter.route(
-  '/receipt/:formId([a-fA-F0-9]{24})/:submissionId([a-fA-F0-9]{24})',
-).get(StripeController.getPaymentReceipt)
+  '/receipt/:formId([a-fA-F0-9]{24})/:submissionId([a-fA-F0-9]{24})/download',
```
Let's add a TODO comment here that we should consider rate limiting this endpoint 🤔 . Please create an issue for that as well.

wanlingt replied:
[c8ea1be](https://github.com/opengovsg/FormSG/pull/5837/commits/c8ea1bead9d14ab779ecdda23d00f62ca154993c)

issue #5924

timotheeg replied:
Thanks @wanlingt , I added some rationale in the ticket.

wanlingt committed:
fix: clarify comment

wanlingt committed:
fix: move tracer to generatePdfFromHtml function

wanlingt committed:
fix: add TODO for rate limiting

timotheeg commented on src/app/services/mail/mail.utils.ts:
```diff
@@ -174,7 +149,7 @@ export const generateAutoreplyPdf = (
   return tracer.trace('generateAutoreplyPdf', () =>
```
Now that `generatePdfFromHtml()` does its own span, we can probably remove this `span` 🤔 (because it doesn't wrap any other action than the pdf generation)

wanlingt replied:
ooh yes thanks for spotting that tim. removed! [d662ee0](https://github.com/opengovsg/FormSG/pull/5837/commits/d662ee0b245d096eb0dd00d1b8522df39a47987d)

timotheeg reviewed (approved):
LGTM 👍 

I added one small comment on the DD span, but other than that, I think it's ready to go! 🎉

tshuli commented:
lgtm, thanks @wanlingt for taking this up!

only one minor thing which I considered during the review - currently we return the pdf file directly in the api response, and trigger the download using `Content-Disposition: attachment` header, and therefore we can't detect when the download is started or completed. This is because the browser handles the download process directly and doesn't provide any hooks or events to notify us when the download is complete.

As a result, we can't display the loading states in the save button during the 2-3s it takes for us to generate the pdf, or handle errors if the download fails (e.g. our server fails to fetch receipt from stripe, we still display the toast with  download started).
![Recording 2023-03-14 at 14 00 33](https://user-images.githubusercontent.com/63710093/224910604-ff8c58b4-83ff-4e2c-8803-c70a676da272.gif)

Perhaps a more complete implementation might indeed be to store the generated pdf in s3, and return the pre-signed url in the api call for the browser to navigate to. This would allow us to properly handle all the states leading up to the start of the file download. But we discussed that we would not do that for this PR, so as to keep things simple.

This note is just for future reference, if this issue ever comes up again.

wanlingt committed:
fix: remove generateAutoreplyPdf span
